client/lib/create-cached-selector/index.js

```diff
@@ -0,0 +1,102 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import { isObject, forEach, some, isFunction } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import warn from 'lib/warn';
+
+/*
+ * A map that is Weak with objects but Strong with primitives
+ */
+export class MixedMap {
+	weakMap = new WeakMap();
+	map = new Map();
+
+	constructor( init ) {
+		forEach( init, ( [ key, val ] ) => {
+			this.mapForKey( key ).set( key, val );
+		} );
+	}
+
+	mapForKey = key => ( isObject( key ) ? this.weakMap : this.map );
+
+	clear() {
+		this.weakMap = new WeakMap();
+		this.map.clear();
+		return this;
+	}
+
+	set( k, v ) {
+		this.mapForKey( k ).set( k, v );
+		return this;
+	}
+
+	delete( k ) {
+		this.mapForKey( k ).delete( k );
+		return this;
+	}
+
+	get( k ) {
+		return this.mapForKey( k ).get( k );
+	}
+
+	has( k ) {
+		return this.mapForKey( k ).has( k );
+	}
+}
+
+/**
+ * Returns a selector that caches values.
+ *
+ * @param  {Function} selector      A standard selector for calculating cached result
+ * @param  {Function} getDependents A Function describing the dependent(s) of the selector.
+ *                                    Must return an object which gets used as the first arg of the selector
+ * @return {Function}               Cached selector
+ */
+export default function createCachedSelector( { selector, getDependents } ) {
+	if ( ! isFunction( selector ) || ! isFunction( getDependents ) ) {
+		throw new TypeError(
+			'createCachedSelector: invalid arguments passed, selector and getDependents must both be functions'
+		);
+	}
+
+	const cache = new MixedMap();
+
+	const cachedSelector = function( state, ...args ) {
+		const dependents = getDependents( state, ...args );
+		const sortedDependentsArray = Object.keys( dependents )
+			.sort()
+			.map( key => dependents[ key ] );
+
+		if ( process.env.NODE_ENV !== 'production' ) {
+			if ( some( args, isObject ) ) {
+				warn( 'Do not pass complex objects as arguments to a cachedSelector' );
```

Comment:
why is this important? we have a number of existing selectors which take objects as arguments after `state`

Reply:
It's part of the current createSelector as well. 

If an argument is an object literal then it shouldn't be used for caching because it's a new object on each call. if it is in object from state, then it should be part of getDependents

Reply:
> because it's a new object on each call

assumption

still, it may be part of `createSelector`, but what I understand is that we're saying, "if you pass any object into a selector we won't let you cache things" when it seems like it could be valid

like what if the dependents are based on a property in that object, or unrelated to them at all?

I think this might occur when we pass things like formatting info to the selectors and stuff like that where the last step in the selector is to do some specific twiddling based on the passed option bag

Reply:
> assumption

object literals are _by definition_ new every time used (like the options bag you suggested).

----

The reason we can't use them is because there is no standard way to create a unique for objects. for an ordered list of primitives its simple (`args.join()`).

If this ends up being a need then we can allow for custom cacheKey generation...but i'd rather stay away for this first iteration because YAGNI

Reply:
can drop the "complex" part as even the empty object qualifies here. Maybe "You can only pass primitives as arguments to a cached selector"

Reply:
dropped the word complex in the latest update.